### PR TITLE
connection handler: remove dynamic casts

### PR DESCRIPTION
**Pull request by mattklein123**
And some other misc cleanup.

Prep for finishing basic UDP proxying.

Risk Level: Low
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@danzh2010 PTAL when you get a chance.

File: source/server/connection_handler_impl.h [hunk 1]
```diff
@@ -227,14 +227,18 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
     Event::TimerPtr timer_;
   };
 
-  Network::ConnectionHandler::ActiveListener*
-  findActiveListenerByAddress(const Network::Address::Instance& address);
+  using ActiveTcpListenerOptRef = absl::optional<std::reference_wrapper<ActiveTcpListener>>;
+
+  struct ActiveListenerDetails {
```
Comment (danzh2010) [inline source/server/connection_handler_impl.h]: nit: comment when the optional tcp_listener_ is empty.

File: source/server/connection_handler_impl.cc [hunk 2]
```diff
@@ -356,12 +353,15 @@ void ConnectionHandlerImpl::ActiveTcpListener::post(Network::ConnectionSocketPtr
     // TODO(mattklein123): We should probably use a hash table here to lookup the tag instead of
     // iterating through the listener list.
     for (const auto& listener : parent.listeners_) {
-      if (listener.second->listener() != nullptr && listener.second->listenerTag() == tag) {
-        // TODO(mattklein123): Remove dynamic_cast here in follow up cleanup.
-        ActiveTcpListener& tcp_listener = *dynamic_cast<ActiveTcpListener*>(listener.second.get());
-        tcp_listener.onAcceptWorker(std::move(socket_to_rebalance->socket),
-                                    tcp_listener.config_.handOffRestoredDestinationConnections(),
-                                    true);
+      if (listener.second.tcp_listener_.has_value() &&
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/server/connection_handler_impl.cc]: Yup good call, will fix.

**Comment by mattklein123** [comment]
@danzh2010 updated

**Pull request merged**